rgdps/api/__init__.py

```diff
@@ -205,12 +224,62 @@ def init_routers(app: FastAPI) -> None:
 def init_middlewares(app: FastAPI) -> None:
     @app.middleware("http")
     async def mysql_transaction(request: Request, call_next):
+        logger.debug(
+            "Opened a new MySQL transaction for request.",
+            extra={
+                "uuid": request.state.uuid,
+            },
+        )
         async with app.state.mysql.transaction() as sql:
             request.state.mysql = sql
             return await call_next(request)
 
+    @app.middleware("http")
+    async def enforce_user_agent(
+        request: Request,
+        call_next: RequestResponseEndpoint,
+    ) -> Response:
+        # Verifying request header for client endpoints.
+        if str(request.url).startswith(config.http_url_prefix):
+            # GD sends an empty User-Agent header.
+            if request.headers.get("User-Agent") != "":
+                logger.info(
+                    "Client request stopped due to invalid User-Agent header.",
+                    extra={
+                        "url": str(request.url),
+                        "uuid": request.state.uuid,
+                    },
+                )
+                return Response(str(GenericResponse.FAIL))
+
+        return await call_next(request)
+
+    @app.middleware("http")
+    async def exception_logging(
+        request: Request,
+        call_next: RequestResponseEndpoint,
+    ) -> Response:
+        try:
+            return await call_next(request)
+        except Exception as e:
+            logger.error(
```

Comment:
use logger.exception() for cases where u wish for the exception to be included in the log

Reply:
^ this is the same as manually passing `exc_info=e` in `logger.error()`